codegen/src/main/ops/org/nd4j/codegen/ops/SDBaseOps.kt

```diff
@@ -0,0 +1,1806 @@
+                Product array reduction operation, optionally along specified dimensions
+            """.trimIndent()
+        }
+    }
+
+    Op("prod") {
+        javaPackage = "org.nd4j.linalg.api.ops.impl.reduce.same"
+        legacy = true
+        Input(NUMERIC, "x") { description = "Input variable" }
+        Arg(BOOL, "keepDims") { description = "If true: keep the dimensions that are reduced on (as size 1). False: remove the reduction dimensions" }
+        Arg(INT, "dimensions") { count = AtLeast(1);  description = "" }
+        Output(NUMERIC, "output"){ description = "" }
+        Doc(Language.ANY, DocScope.ALL){
+            """
+                Product array reduction operation, optionally along specified dimensions
+                Note that if keepDims = true, the output variable has the same rank as the input variable,
+                with the reduced dimensions having size 1. This can be useful for later broadcast operations (such as subtracting
+                the mean along a dimension).
+                Example: if input has shape [a,b,c] and dimensions=[1] then output has shape:
+                keepDims = true: [a,1,c]
+                keepDims = false: [a,c]
+            """.trimIndent()
+        }
+    }
+
+    Op("range") {
+        javaPackage = "org.nd4j.linalg.api.ops.random.impl"
+        Arg(NUMERIC, "from") { description = "Initial/smallest value" }
+        Arg(NUMERIC, "to") { description = "Largest value (exclusive)" }
+        Arg(NUMERIC, "step") { description = "Step size" }
+        Arg(DATA_TYPE, "dataType") { description = "" }
+        Output(NUMERIC, "output"){ description = "SDVariable with the specified values" }
+
+        Doc(Language.ANY, DocScope.ALL){
+            """
+                Create a new variable with a 1d array, where the values start at {@code from} and increment by {@code step}
+                up to (but not including) limit.
+                For example, {@code range(1.0, 3.0, 0.5)} will return {@code [1.0, 1.5, 2.0, 2.5]
+            """.trimIndent()
+        }
+    }
+
+    Op("rank") {
+        javaPackage = "org.nd4j.linalg.api.ops.impl.shape"
+        Input(NUMERIC, "in") { description = "Input variable" }
+        Output(NUMERIC, "output"){ description = "(scalar) output variable with value equal to the rank of the input variable" }
+        Doc(Language.ANY, DocScope.ALL){
+            """
+                Returns the rank (number of dimensions, i.e., length(shape)) of the specified SDVariable as a 0D scalar variable
+            """.trimIndent()
+        }
+    }
+
+    Op("repeat") {
+        javaPackage = "org.nd4j.linalg.api.ops.impl.shape"
+        Input(NUMERIC, "df") {count = AtLeast(1); description = "" }
+        Arg(INT, "axis") { description = "" }
+        Output(NUMERIC, "output"){ description = "" }
+        Doc(Language.ANY, DocScope.ALL){
+            """
+                @see #repeat(String, SDVariable, int)
+            """.trimIndent()
+        }
+    }
```

Comment:
This looks wrong in SameDiff/SDBaseOps also.... need to look into this op generally.